### DISCUSSION TITLE
static spin correlation function for 1d heisenberg hamiltonian

### DISCUSSION THREAD

**Discussion by 000sadaf** [QuSpin/QuSpin#486] [Q&A] [answered]
hello,
i want to make the spin- spin correlation function for heisenberg hamiltonian in ground state. then plot between correlation function (C(r)) and 'r'(number of sites).
_

i did some modification in auto correlation code which is given in the documentation but did not get the result,
the einsum is not working it gives me error. 
so i decided to do for np.dot()
but i did not get the desired result.

and the other thing is that with the periodic boundary condition, if number of spin is N=12, the correlation function for spin 0 and for spin 11 should be same.
but it is not.

please give me some suggation.  here is my code that i modified i followed the autocorrelation function codes which is given 

> .

import quspin as qp
from quspin.operators import hamiltonian
from quspin.basis import spin_basis_1d
import numpy as np
import sympy as sp
import matplotlib.pyplot as plt

L = 12
Jxy = 1.0  #spin interaction
Jzz = 1.0 # z interaction term
distance = np.linspace(0.0,10.0,150)

basis = spin_basis_1d(L, pauli= False, pblock=1)

jzz = [[Jzz, i, (i+1)%L]for i in range(L)]
jxy = [[Jxy/2, i, (i+1)%L]for i in range(L)]

distance = np.linspace(0.0,5.0,50)

#static
static = [["+-", jxy], ["-+", jxy],["zz", jzz]]
dynamic = []
H_XXZ = hamiltonian(static, dynamic, basis=basis, dtype=np.float64)
print(H_XXZ.toarray())

E,V = H_XXZ.eigsh(k=1,which="SA")
psi_GS = V[:,0]

psi_GS = V[:,0]
op_list = [["z",[0],np.sqrt(2.0)]]

Opsi_GS = basis.inplace_Op(psi_GS,op_list,np.float64)

oplist_r = [["z",[r],np.sqrt(2.0)]for r in range(L)]
Opsi_r_GS = basis.inplace_Op(psi_GS,oplist_r,np.float64)
def correlator_(L,distance,S="1/2"):
    C_r = np.einsum("i,i->i",Opsi_GS.conj(),Opsi_r_GS)
    return C_r
Crr = correlator_(L,distance)
plt.plot(distance,Crr.real,'-b',label='no symm.: $\\mathrm{Re}\\;C(r)$')
plt.plot(distance,Crr.imag,'-r',label='no symm.: $\\mathrm{Im}\\;C(r)$')
plt.xlabel('dist $r$')
plt.legend()
#
plt.show()

**Comment by JiahaoYao** [comment]
Hi, I am afraid u might provide MWE in order to specific the question, thanks!

**Reply by 000sadaf** [reply]
i have write down the code .
please give me some suggation.

**Reply by JiahaoYao** [reply]
The current code gives the error 

```python
  File "python3.7/site-packages/matplotlib/axes/_axes.py", line 1605, in plot
    lines = [*self._get_lines(*args, data=data, **kwargs)]
  File "python3.7/site-packages/matplotlib/axes/_base.py", line 315, in __call__
    yield from self._plot_args(this, kwargs)
  File "python3.7/site-packages/matplotlib/axes/_base.py", line 501, in _plot_args
    raise ValueError(f"x and y must have same first dimension, but "
ValueError: x and y must have same first dimension, but have shapes (50,) and (2080,)
```

This is because the shape 
```python
> Opsi_r_GS.shape
(2080,)
> distance.shape
(50,)
```

Do u also have the same error?

**Reply by JiahaoYao** [reply]
Are u trying to change this [code](https://weinbe58.github.io/QuSpin/examples/example19.html)?

**Reply by 000sadaf** [reply]
yes i am facing the same error with this code. now i change the code like this....
psi_GS = V[:,0]
Sz0_list = [["z",[0],1]]
# use inplace_Op to apply operator O on psi_GS
Sz0_GS = basis.inplace_Op(psi_GS,Sz0_list,np.float64)

for r in range(L):
    print(r)
    Szr_list = [["z",[r],1]]
    Szr_GS = basis.inplace_Op(psi_GS,Szr_list,np.float64)
    corr = np.dot(Sz0_GS.conj(),Szr_GS)
    print(corr)

still i could not plot between correlation function and distance(no of sites).
there is always error.
when i print this corr the values i am getting i think is wrong because my hamiltonian is periodic. S(0)=S(L+1)
so the correlation function C(0) and C(L+1) should be same.

**Reply by 000sadaf** [reply]
every time when i run this code to get ground state ..i got different eigenvector and eigen value means eigen value change sometimes but eigen vector change hence corresponding correlation function value change `
E,V = H.eigsh(k=1,which="SA")
why this value fluctuating on every time when i excute

**Reply by JiahaoYao** [reply]
Hi @000sadaf , thanks! One quick question is whether you miss the terms `psi_GS_t` and `Opsi_GS_t` at line 44 and 52 [here](https://weinbe58.github.io/QuSpin/examples/example19.html). 

In the example, there is extra `time` variables; seems in ur code there is no such time. Instead, you have the `distance`. Seems like you want to compute the correlation about the distance, however, in ur function  `correlator_(L,distance,S="1/2")`, the `distance` hasn't been used. Is there anything missed?

**Reply by JiahaoYao** [reply]
> every time when i run this code to get ground state ..i got different eigenvector and eigen value means eigen value change sometimes but eigen vector change hence corresponding correlation function value change ` E,V = H.eigsh(k=1,which="SA") why this value fluctuating on every time when i excute

by the way, everytime I run your code, I can get the same answers. Is this what you mean? 

```python
0
0.2500000000000001
1
-0.1496497477068116
2
0.06261092842265799
3
-0.05527391487113138
4
0.04028491119194629
5
-0.040803496609567697
6
0.03566263914581162
7
-0.040803496609567474
8
0.04028491119194616
9
-0.055273914871131
10
0.06261092842265793
11
-0.14964974770681094
```

**Reply by JiahaoYao** [reply]
> yes i am facing the same error with this code. now i change the code like this.... psi_GS = V[:,0] Sz0_list = [["z",[0],1]]
> 
> # use inplace_Op to apply operator O on psi_GS
> Sz0_GS = basis.inplace_Op(psi_GS,Sz0_list,np.float64)
> 
> for r in range(L): print(r) Szr_list = [["z",[r],1]] Szr_GS = basis.inplace_Op(psi_GS,Szr_list,np.float64) corr = np.dot(Sz0_GS.conj(),Szr_GS) print(corr)
> 
> still i could not plot between correlation function and distance(no of sites). there is always error. when i print this corr the values i am getting i think is wrong because my hamiltonian is periodic. S(0)=S(L+1) so the correlation function C(0) and C(L+1) should be same.

For the symmetry parts, I guess in your code, you choose your `Szr_list = [["z",[r],1]]`; then the symmetry is around the `0` index, that is because you have `Sz0_list = [["z",[0],1]]`. That being said the whole system size still periodic and thus if the index is symmetric about `0`, then the outcome from your code gives the same answers. 

Or another question is whether you want to have `[["z",[r],np.sqrt(2.0)]for r in range(L)]`.

**Reply by 000sadaf** [reply]
``hello thanks  for replying.. actually i am new to coding so facing a problem, i modified the codes so far i am gud with the result .
i got the values of correlation function SzSz , SxSx ,SySy ,S+S- and S-S+. from the commutation relation , correlation func of [SxSx,SySy] =0 and also[S+S- -S-S+]=0.  So far i am getting the good result .BUT
and we know that also corr. fun. SxSx+SySy = 1/2(S+S- + S-S+).
HERE IS AM NOT GETTING THIS.
i am providing my code please you please check and tell me why it is happening. 
THANK YOU..

from quspin.basis import spin_basis_general
from quspin.operators import hamiltonian
from quspin.tools.evolution import expm_multiply_parallel
import numpy as np
import matplotlib.pyplot as plt # plotting library
from scipy.interpolate import make_interp_spline


L = 12
Jxy = 1.0  #spin interaction
Jzz = 0.5 # z interaction term
jzz = [[Jzz, i, (i+1)%L]for i in range(L)]
jxy = [[Jxy/2, i, (i+1)%L]for i in range(L)]

static = [["+-", jxy], ["-+", jxy],["zz", jzz]]

basis = spin_basis_general(L,S="1/2",pauli=False)
# define Heisenberg Hamiltonian 
no_checks = dict(check_symm=False,check_herm=False,check_pcon=False)
H = hamiltonian(static,[],basis=basis,dtype=np.float64,**no_checks)
#print(H.toarray())
E,V = H.eigh()
#print(basis)
#print(E,V)
E,V = H.eigsh(k=1,which="SA")
psi_GS = V[:,0]
#print(psi_GS)
print(E)

Sz0_list = [["z",[0],1]]
Sz0_GS = basis.inplace_Op(psi_GS,Sz0_list,np.float64)

list=[]
list1=[]
for r in range(1,(L)):
    #print(r)
    Szr_list = [["z",[r],1]]
    Szr_GS = basis.inplace_Op(psi_GS,Szr_list,np.float64)
    corr = np.dot(Sz0_GS.conj(),Szr_GS)
    corrSz = np.absolute(corr)
    list1.append(corrSz)
    list.append(r)
    print(corrSz)
#print(list1)
np.shape(list)
x=np.array(list)
y=np.array(list1)
#print(x)
X_Y_Spline = make_interp_spline(x,y)
X_ = np.linspace(x.min(), x.max(),100)
Y_ =  X_Y_Spline(X_)
plt.plot(X_,Y_)
plt.xlabel("Number of sites (r)")
plt.ylabel("Correlation")
plt.grid()

Sx0_list = [["x",[0],1]]
# use inplace_Op to apply operator O on psi_GS
Sx0_GS = basis.inplace_Op(psi_GS,Sx0_list,np.float64)

list2=[]
list3=[]
for j in range(1,(L)):
    #print(j)
    Sxr_list = [["x",[j],1]]
    Sxr_GS = basis.inplace_Op(psi_GS,Sxr_list,np.float64)
    corr1 = np.dot(Sx0_GS.conj(),Sxr_GS)
    corrSx = np.absolute(corr1)
    list3.append(corrSx)
    list2.append(j)
    print(corrSx)
#print(list2)
np.shape(list)
a=np.array(list2)
b=np.array(list3)
print(a)
X_Y_Spline1 = make_interp_spline(a,b)
X_1 = np.linspace(a.min(), a.max(),5000)
Y_1 =  X_Y_Spline1(X_1)
plt.plot(X_1,Y_1)
plt.xlabel("Number of sites (j)")
plt.ylabel("Correlation(SxSx)")
plt.grid()

Sp0_list = [["+",[0],1]]
# use inplace_Op to apply operator O on psi_GS
Sp0_GS = basis.inplace_Op(psi_GS,Sp0_list,np.float64)

list4=[]
list5=[]
for k in range(1,(L)):
    #print(k)
    Snr_list = [["-",[k],1]]
    Snr_GS = basis.inplace_Op(psi_GS,Snr_list,np.float64)
    corr2 = np.dot(Sp0_GS.conj(),Snr_GS)
    corrSpn = np.absolute(corr2)
    list5.append(corrSpn)
    list4.append(k)
    print(corr2)
#print(list4)

c=np.array(list5)
d=np.array(list4)
print(d)
X_Y_Spline2 = make_interp_spline(d,c)
X_2 = np.linspace(d.min(), d.max(),1000)
Y_2 =  X_Y_Spline2(X_2)
plt.plot(X_2,Y_2)
plt.xlabel("Number of sites (k)")
plt.ylabel("Correlation(SpSn)")
plt.grid()
Sn0_list = [["-",[0],1]]
# use inplace_Op to apply operator O on psi_GS
Sn0_GS = basis.inplace_Op(psi_GS,Sn0_list,np.float64)

list6=[]
list7=[]
for i in range(1,(L)):
    #print(j)
    Spr_list = [["+",[i],1]]
    Spr_GS = basis.inplace_Op(psi_GS,Spr_list,np.float64)
    corr3 = np.dot(Sn0_GS.conj(),Spr_GS)
    corrSnp = np.absolute(corr3)
    list7.append(corrSnp)
    list6.append(i)
    print(corr3)
#print(list6)

e=np.array(list6)
f=np.array(list7)
print(e)
X_Y_Spline3 = make_interp_spline(e,f)
X_3 = np.linspace(e.min(), e.max(),500)
Y_3 =  X_Y_Spline3(X_3)
plt.plot(X_3,Y_3)
plt.xlabel("Number of sites (i)")
plt.ylabel("Correlation(SnSp)")
plt.grid()
Sy0_list = [["y",[0],1]]
# use inplace_Op to apply operator O on psi_GS
Sy0_GS = basis.inplace_Op(psi_GS,Sy0_list,np.complex128)

list8=[]
list9=[]
for p in range(1,(L)):
    #print(p)
    Syr_list = [["y",[p],1]]
    Syr_GS = basis.inplace_Op(psi_GS,Syr_list,np.complex128)
    corr4 = np.dot(Sy0_GS.conj(),Syr_GS)
    corrSy = np.absolute(corr4)
    list9.append(corrSy)
    list8.append(p)
    print(corrSy)
#print(list6)

g=np.array(list8)
h=np.array(list9)
print(g)
X_Y_Spline4 = make_interp_spline(g,h)
X_4 = np.linspace(g.min(), g.max(),500)
Y_4 =  X_Y_Spline4(X_4)
plt.plot(X_4,Y_4)
plt.xlabel("Number of sites (p)")
plt.ylabel("Correlation(SySy)")
plt.grid()

totSxSy = corrSx+corrSy
totSpnSnp = (corrSpn+corrSnp)/2
rat = totSxSy/totSpnSnp
print(rat)
print(totSxSy)
print(totSpnSnp)

**Comment by 000sadaf** [comment, marked as answer]
hello ,please let me know the answer why it is happening, that 1/2corr(S+S- + S-S+) is not equal to corr(SxSx + SySy).

**Reply by weinbe58** [reply]
Can you attach the full code as a text file? Thank you!

**Reply by weinbe58** [reply]
Actually I know what the issue is, its related to how you're calculating the expectation value at the end. with S+ and S- you're assuming that <0|S+ = (S+|0>)^T which is not true, it should be <0|S+ = (S-|0>)^T

**Reply by weinbe58** [reply]
here is a working example where you need to add the argument `transpose=True` and `conjugated=True` because this acts with the operator from the right not from the left.

```python
from quspin.basis import spin_basis_general
from quspin.operators import hamiltonian
from quspin.tools.evolution import expm_multiply_parallel
import numpy as np
import matplotlib.pyplot as plt # plotting library
from scipy.interpolate import make_interp_spline



def Corr(basis,psi,a,b):
	# <psi|ab|psi>
	N = basis.N
	C = np.zeros(N)

	op_list = [[a,[0],1.0]]
	O0_psi_GS = basis.inplace_Op(psi_GS,op_list,np.complex128)

	for r in range(N):
		op_list = [[b,[r],1.0]]
		Or_psi_GS = basis.inplace_Op(psi_GS,op_list,np.complex128,transposed=True,conjugated=True)

		C[r] = np.vdot(Or_psi_GS,O0_psi_GS).real

	return C


L = 12
Jxy = 1.0 #spin interaction
Jzz = 0.5 # z interaction term
jzz = [[Jzz, i, (i+1)%L]for i in range(L)]
jxy = [[Jxy/2, i, (i+1)%L]for i in range(L)]

static = [["+-", jxy], ["-+", jxy],["zz", jzz]]

basis = spin_basis_general(L,S="1/2",pauli=False)

no_checks = dict(check_symm=False,check_herm=False,check_pcon=False)
H = hamiltonian(static,[],basis=basis,dtype=np.float64,**no_checks)

E,V = H.eigsh(k=1,which="SA")
psi_GS = V[:,0]
#print(psi_GS)
print(E)



SpSm = Corr(basis,psi_GS,"+","-") # 
SmSp = Corr(basis,psi_GS,"-","+")
SzSz = Corr(basis,psi_GS,"z","z")
SySy = Corr(basis,psi_GS,"y","y")
SxSx = Corr(basis,psi_GS,"x","x")

print(0.5*(SpSm+SmSp) - (SxSx+SySy))
plt.plot(0.5*(SpSm+SmSp))
plt.plot((SxSx+SySy))

plt.show()




```

**Reply by 000sadaf** [reply]
Thank you so much. this was very minor thing i should have known that.

**Reply by thefacemyer** [reply]
Why does this not work for higher spin? I just plugged in L=4 and S="3/2", and it says the operator is not recognized.

**Reply by JiahaoYao** [reply]
Hi @thefacemyer , this seems like a bug! would you please open a ticket? (in case this gets lost) Thanks!
